main.go

```diff
@@ -46,68 +49,76 @@ func main() {
 
 	if *cloud == "" && !*multiCloud {
-		log.Fatalln("openstack-exporter: error: required argument 'cloud' or flag --multi-cloud not provided, try --help")
-	}
-	err := log.Base().SetLevel(*logLevel)
-	if err != nil {
-		log.Errorf("Cannot init set logger level: %s", err)
-		os.Exit(-1)
+		level.Error(logger).Log("msg", "openstack-exporter: error: required argument 'cloud' or flag --multi-cloud not provided, try --help")
 	}
 
-	log.Infoln("Build context", version.BuildContext())
+	level.Info(logger).Log("msg", "Build context", "build_context", version.BuildContext())
 
 	if *osClientConfig != DEFAULT_OS_CLIENT_CONFIG {
-		log.Debugf("Setting Env var OS_CLIENT_CONFIG_FILE = %s", *osClientConfig)
+		level.Debug(logger).Log("msg", "Setting Env var OS_CLIENT_CONFIG_FILE", "os_client_config_file", *osClientConfig)
 		os.Setenv("OS_CLIENT_CONFIG_FILE", *osClientConfig)
 	}
 
-	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
-		_, err := w.Write([]byte(`<html>
-             <head><title>OpenStack Exporter</title></head>
-             <body>
-             <h1>OpenStack Exporter</h1>
-             <p><a href='` + *metrics + `'>Metrics</a></p>
-             </body>
-             </html>`))
-		if err != nil {
-			log.Error(err)
-		}
-	})
+	links := []web.LandingLinks{}
+
 	if *multiCloud {
-		http.HandleFunc("/probe", probeHandler(services))
-		http.Handle("/metrics", promhttp.Handler())
-		log.Infoln("openstack exporter started in multi cloud mode (/probe?cloud=)")
+		http.HandleFunc("/probe", probeHandler(services, logger))
+		http.Handle(*metrics, promhttp.Handler())
+		level.Info(logger).Log("msg", "openstack exporter started in multi cloud mode (/probe?cloud=)")
+		links = append(links, web.LandingLinks{
+			Address: *metrics,
+			Text:    "Metrics",
+		}, web.LandingLinks{
+			Address: "/probe",
+			Text:    "Probes",
+		})
 	} else {
-		log.Infoln("openstack exporter started in legacy mode")
-		http.HandleFunc(*metrics, metricHandler(services))
+		level.Info(logger).Log("msg", "openstack exporter started in legacy mode")
+		http.HandleFunc(*metrics, metricHandler(services, logger))
+		links = append(links, web.LandingLinks{
+			Address: *metrics,
+			Text:    "Metrics",
+		})
 	}
 
-	if *bind == "" {
-		log.Info("--web.listen-address is empty. HTTP server will start on :9180")
-		*bind = ":9180"
+	if *metrics != "/" && *metrics != "" {
+		landingConfig := web.LandingConfig{
+			Name:        "openstack_exporter",
+			Description: "Prometheus Exporter for openstack",
+			Version:     version.Info(),
+			Links:       links,
+		}
+
+		landingPage, err := web.NewLandingPage(landingConfig)
+		if err != nil {
+			level.Error(logger).Log("err", err)
+			os.Exit(1)
+		}
+		http.Handle("/", landingPage)
 	}
 
 	if *domainID != "" {
-		log.Infoln("Gathering metrics for domain ID", *domainID)
+		level.Info(logger).Log("msg", "Gathering metrics for configured domain ID", "domain_id", *domainID)
 	}
 
-	tcp := exporters.IP4or6(*bind)
```

Comment:
why are you removing this?

Reply:
The server as well as the config `web.listen-address` is handled by the [exporter-toolkit](github.com/prometheus/exporter-toolkit) [here](https://github.com/prometheus/exporter-toolkit/blob/e26aafa966f841a37eb9fcd91cc7ea8759e00225/web/tls_config.go#L300). It is using the default of the `IP4or6` method `tcp`. So there is no need for us to set the listen socket to a specified protocol. Hope it explains why it is removed.